### PR TITLE
#13[feat/money-stream] Vesting/Payment stream using Streamflow SDK

### LINKED ISSUE #13: [feat/money-stream] Implement Money Streaming using Streamflow SDK

### PR DESCRIPTION
**Pull request by TropicalDog17**
test query: `I'm looking to make a 1 SOL token stream payment every second for 3 month to the recipent 4WMjxRZ1HhX4RhZ1fiohpwUTmjeCudQhwYogzvqHKSjh`

### CODE REVIEW

**Comment by TropicalDog17** [comment]
@batm9n Kindly try to run this draft PR and check if it works for you.

**Comment by MrSufferer** [comment]
@TropicalDog17 Sure, let's check it out.

**1_ First signed message _**

Screen Shot 2023-10-09 at 21 09 16



Updating...

**Comment by TropicalDog17** [comment]
@batm9n To my knowledge it is a problem related to Solana Associated Token Account, you need to create and fund first.

+ (if already created but no fund): `spl-token unwrap`
+ Then we can fund the ATA with Wrapped Sol: `spl-token wrap 10`

Another problem is that if the recipent doesn't have an ATA, the streaming tx will fail as well. So I have tried to create ATA account for the recipent on the fly, but the error is quite unexpected when getOrCreateAssociatedTokenAccount() is evoked.

I guess we need to handle the error, because although it throws an error, an ATA is still created and worked smoothly the second time we run.

**Comment by MrSufferer** [comment]
@TropicalDog17
1. It would be better if we could manage to auto-convert SOL->WSOL for users.
2. getOrCreateAssociatedTokenAccount() might return an account that hasn't reached the "confirmation state".
- Maybe this sample can help you:  https://github.com/batm9n/plink/blob/main/ui_components/ShareLinkPage.tsx#L316-L365

**Comment by MrSufferer** [comment]
@TropicalDog17 
Have you added Wrapped Sol convert yet?
Reference: https://solanacookbook.com/references/token.html#how-to-manage-wrapped-sol

File: handler/.prettierrc
Comment (MrSufferer) [inline handler/.prettierrc:1]: lack an endline

File: handler/utils/streamFlow.js
Comment (MrSufferer) [inline handler/utils/streamFlow.js:1]: Line 21: What is canTopUp?

File: handler/test.js
Comment (MrSufferer) [inline handler/test.js:1]: what is the kind of test?

File: handler/transpiler.js
Comment (MrSufferer) [inline handler/transpiler.js:1]: line 180: `nft`-> `stream`

File: handler/data/trainingData.js
Comment (MrSufferer) [inline handler/data/trainingData.js:1]: - Line 1015: don't hardcode the stream recipient address

Comment (TropicalDog17) [reply handler/data/trainingData.js:1]: Can you clarify on this? I think this address only works as a placeholder address, and  I don't see any problem

File: handler/test.js
Comment (TropicalDog17) [reply handler/test.js:1]: Sorry, just my dummy file for testing. I'll remove it.

File: handler/utils/streamFlow.js
Comment (TropicalDog17) [reply handler/utils/streamFlow.js:1]: It's a stream param: Read more here: https://github.com/streamflow-finance/js-sdk#create-stream
`  canTopup: false, // setting to FALSE will effectively create a vesting contract.`

File: handler/data/trainingData.js
Comment (MrSufferer) [reply handler/data/trainingData.js:1]: ok it's fine

**Comment by TropicalDog17** [comment]
> @TropicalDog17 Have you added Wrapped Sol convert yet? Reference: https://solanacookbook.com/references/token.html#how-to-manage-wrapped-sol

@batm9n Done, sorry for the late fix, was struggling a lot with this task.

**Review by MrSufferer — COMMENTED** [review]
@TropicalDog17 Don't worry, Solana is a struggle itself.

Here is my experience with the latest commit:

1. My first try
Screen Shot 2023-10-11 at 22 48 32

You haven't handled the extracted amount (I requested 0.01 SOL but it ended up sending 1 SOL)

*_*Signing:*_*
Screen Shot 2023-10-11 at 22 44 08


2. My second try:

Screen Shot 2023-10-11 at 22 48 11

**Comment by MrSufferer** [comment]
@TropicalDog17
I've commented out the dummy `streamInfo`, then tried with Devnet and it worked.
It seems ok.

A few notes:
- Did you automatically convert 1 SOL to WSOL? Can you leave it with an arbitrary amount?
- I didn't see my subtracted SOL when signing with the wallet. Could you provide an explorer URL where the stream is conducted?

**Comment by TropicalDog17** [comment]
> @TropicalDog17 I've commented out the dummy `streamInfo`, then tried with Devnet and it worked. It seems ok.

@batm9n  yeah my apologize, I forgot to mention the dummy part.

**Comment by MrSufferer** [comment]
@TropicalDog17 nice. Push it and I will merge soon.

**Comment by TropicalDog17** [comment]
@batm9n 

> @TropicalDog17 I've commented out the dummy `streamInfo`, then tried with Devnet and it worked. It seems ok.
> 
> A few notes:
> 
>     * Did you automatically convert 1 SOL to WSOL? Can you leave it with an arbitrary amount?
> 
>     * I didn't see my subtracted SOL when signing with the wallet. Could you provide an explorer URL where the stream is conducted?

1. The conversion is based on the amount of SOL we need to send plus streamflow fees. I know it's still bad(maybe considering the sender balance is a good idea), but yeah.

2. I have console.log-ed the address after the stream creation. You can see an example here: https://solana.fm/address/Gjm4LK6KxVZyu1uyAsjjyAJyweYYgqShqBVLVrXQnMcs?cluster=devnet-solana

You can also login to [streamflow.finance](https://app.streamflow.finance) as devnet environment and see your outgoing/ingoing stream, it does a great job providing statistic of our streams.

**Comment by MrSufferer** [comment]
LGTM!

**Pull request merged**